### PR TITLE
Enable gnn-based promting for gpeft

### PR DESCRIPTION
**Pull request by GentleZhu**
*Issue #, if available:*

*Description of changes:*
Finish the todos in the original peft llm gnn PR (#673).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 1]
```diff
@@ -115,32 +116,56 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
...
         attention_mask = self._lm_node_feats[self.target_ntype][ATT_MASK_IDX][output_nodes].to(self.llm.device)
-        # TODO (qzhuamzn): modify input_ids into input_embeds=[graph_tokens, input_embeds] to support GPEFT
-        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask)
+        graph_tokens = self.gnn_encoder(blocks, node_feats)[self.target_ntype]
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+        word_embeddings = self.llm.get_input_embeddings()
+        # Tentatively leave the zero prompting line for debugging and testing
+        #inputs_embeds = torch.cat([torch.zeros_like(graph_tokens.unsqueeze(1)), word_embeddings(input_ids)], dim=1)
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:134]: remove the commented code.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 2]
```diff
@@ -115,32 +116,56 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
...
         attention_mask = self._lm_node_feats[self.target_ntype][ATT_MASK_IDX][output_nodes].to(self.llm.device)
-        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask)
-        logits = self.out(model_output.last_hidden_state[:,-1,:])
+        graph_tokens = self.gnn_encoder(blocks, node_feats)[self.target_ntype]
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+        word_embeddings = self.llm.get_input_embeddings()
+        # Tentatively leave the zero prompting line for debugging and testing
+        #inputs_embeds = torch.cat([torch.zeros_like(graph_tokens.unsqueeze(1)), word_embeddings(input_ids)], dim=1)
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:160]: the same comment

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 3]
```diff
@@ -115,32 +116,56 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
...
-        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask)
+        graph_tokens = self.gnn_encoder(blocks, node_feats)[self.target_ntype]
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+        word_embeddings = self.llm.get_input_embeddings()
+        # Tentatively leave the zero prompting line for debugging and testing
+        #inputs_embeds = torch.cat([torch.zeros_like(graph_tokens.unsqueeze(1)), word_embeddings(input_ids)], dim=1)
+        inputs_embeds = torch.cat([graph_tokens.unsqueeze(1), word_embeddings(input_ids)], dim=1)
+        attention_mask = torch.cat([torch.ones((input_shape[0],1), device=self.llm.device), attention_mask], dim=1)
```
Comment (classicsong) [inline examples/peft_llm_gnn/llm_gnn_model.py:136]: Shouldn't the mask of graph tokens be graph_tokens.shape[1]?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 4]
```diff
@@ -115,32 +116,64 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
...
         attention_mask = self._lm_node_feats[self.target_ntype][ATT_MASK_IDX][output_nodes].to(self.llm.device)
-        # TODO (qzhuamzn): modify input_ids into input_embeds=[graph_tokens, input_embeds] to support GPEFT
-        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask)
+
+        graph_tokens = self.encode_graph(blocks, node_feats)
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+        word_embeddings = self.llm.get_input_embeddings()
+        inputs_embeds = torch.cat([graph_tokens.unsqueeze(1), word_embeddings(input_ids)], dim=1)
```
Comment (classicsong) [inline examples/peft_llm_gnn/llm_gnn_model.py:144]: Can you add some comments to describe the shape of inputs_embeds and attention_masks and how graph-embeddings are attached into test-embeddings?

**Review by classicsong — APPROVED** [review]
